File: NoteBase/Storage/AppPaths.cs
using System.IO;
using System.Reflection;

namespace NoteBase.Storage
{
    /// <summary>
    /// MemoRoot 配下の各サブパスを提供する。
    /// </summary>
    public class AppPaths
    {
        public string Root { get; private set; }

        public AppPaths(string root)
        {
            Root = root;
        }

        public string Notes { get { return Path.Combine(Root, "notes"); } }
        public string Trash { get { return Path.Combine(Root, "trash"); } }
        public string Outputs { get { return Path.Combine(Root, "outputs"); } }
        public string Indexes { get { return Path.Combine(Root, "indexes"); } }
        public string Templates { get { return Path.Combine(Root, "templates"); } }
        public string Canvases { get { return Path.Combine(Root, "canvases"); } }
        public string Config { get { return Path.Combine(Root, "config"); } }

        public string NoteDir(string id) { return Path.Combine(Notes, id); }
        public string MetaPath(string id) { return Path.Combine(NoteDir(id), "meta.json"); }
        public string IndexMdPath(string id) { return Path.Combine(NoteDir(id), "index.md"); }
        public string ImagesDir(string id) { return Path.Combine(NoteDir(id), "images"); }

        public void EnsureLayout()
        {
            Directory.CreateDirectory(Root);
            Directory.CreateDirectory(Notes);
            Directory.CreateDirectory(Trash);
            Directory.CreateDirectory(Outputs);
            Directory.CreateDirectory(Indexes);
            Directory.CreateDirectory(Templates);
            Directory.CreateDirectory(Canvases);
            Directory.CreateDirectory(Config);
        }

        /// <summary>
        /// 既定の MemoRoot は実行ファイル直下の "MemoRoot/" とする。
        /// </summary>
        public static AppPaths Default()
        {
            var exeDir = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
            return new AppPaths(Path.Combine(exeDir, "MemoRoot"));
        }
    }
}

File: notebase/storage/app_paths.py
from __future__ import annotations

import os
from pathlib import Path


class AppPaths:
    """MemoRoot 配下の各サブパスを提供する。"""

    def __init__(self, root: str | Path) -> None:
        self.root = Path(root)

    @property
    def notes(self) -> Path:
        return self.root / "notes"

    @property
    def trash(self) -> Path:
        return self.root / "trash"

    @property
    def outputs(self) -> Path:
        return self.root / "outputs"

    @property
    def indexes(self) -> Path:
        return self.root / "indexes"

    @property
    def templates(self) -> Path:
        return self.root / "templates"

    @property
    def canvases(self) -> Path:
        return self.root / "canvases"

    @property
    def config(self) -> Path:
        return self.root / "config"

    def note_dir(self, note_id: str) -> Path:
        return self.notes / note_id

    def meta_path(self, note_id: str) -> Path:
        return self.note_dir(note_id) / "meta.json"

    def index_md_path(self, note_id: str) -> Path:
        return self.note_dir(note_id) / "index.md"

    def images_dir(self, note_id: str) -> Path:
        return self.note_dir(note_id) / "images"

    def ensure_layout(self) -> None:
        for p in (
            self.root,
            self.notes,
            self.trash,
            self.outputs,
            self.indexes,
            self.templates,
            self.canvases,
            self.config,
        ):
            p.mkdir(parents=True, exist_ok=True)

    @classmethod
    def default(cls) -> "AppPaths":
        """既定の MemoRoot を返す。

        優先順:
          1. 環境変数 NOTEBASE_ROOT
          2. リポジトリ直下の MemoRoot/  (notebase パッケージの 2 階層上)
        """
        env = os.environ.get("NOTEBASE_ROOT")
        if env:
            return cls(env)
        # __file__ = .../notebase/storage/app_paths.py → 2 階層上がリポジトリルート
        repo_root = Path(__file__).resolve().parent.parent.parent
        return cls(repo_root / "MemoRoot")
